org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java

```diff
@@ -595,9 +597,16 @@ public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
 	@Override
 	public CompletableFuture<Object> executeCommand(ExecuteCommandParams params) {
 		debugTrace(">> workspace/executeCommand " + (params == null ? null : params.getCommand()));
-		return computeAsync((monitor) -> {
-			return commandHandler.executeCommand(params, monitor);
-		});
+		// see https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/2799
+		// Handle the paste event synchronously in the main thread
+		if (IDelegateCommandHandler.JAVA_EDIT_HANDLE_PASTE_EVENT.equals(params.getCommand())) {
+			Object result = commandHandler.executeCommand(params, new NullProgressMonitor());
```

Comment:
Is your thinking here that it's pointless to report progress on the paste event ? I could of agree. I know we don't currently report progress to the client for delegates via progress support but [we may implement it in the future](https://github.com/eclipse-jdtls/eclipse.jdt.ls/issues/1186).